### PR TITLE
Return 404 for non-existent queues for the visibility endpoints

### PR DESCRIPTION
**Pull request by PBundyra**
#### What type of PR is this?
/kind bug

#### Special notes for your reviewer:
Method `cmpopts.EquateErrors()` returned diff due to different addresses of errors.

#### Does this PR introduce a user-facing change?

```release-note
Visibility endpoints return 404 code for non-existent queues
```

### CODE REVIEW

File: pkg/visibility/api/rest/pending_workloads_cq.go [hunk 1]
```diff
@@ -66,6 +67,10 @@ func (m *pendingWorkloadsInCqREST) Get(ctx context.Context, name string, opts ru
 
 	wls := make([]v1alpha1.PendingWorkload, 0, limit)
 	pendingWorkloadsInfo := m.queueMgr.PendingWorkloadsInfo(name)
+	if pendingWorkloadsInfo == nil {
+		return nil, errors.NewNotFound(v1alpha1.Resource("clusterqueue"), name)
```
Comment (mimowo) [inline pkg/visibility/api/rest/pending_workloads_cq.go:71]: Can we have this tested, similarly as for LQs?

File: pkg/visibility/api/rest/pending_workloads_cq.go [hunk 2]
```diff
@@ -66,6 +67,10 @@ func (m *pendingWorkloadsInCqREST) Get(ctx context.Context, name string, opts ru
 
 	wls := make([]v1alpha1.PendingWorkload, 0, limit)
 	pendingWorkloadsInfo := m.queueMgr.PendingWorkloadsInfo(name)
+	if pendingWorkloadsInfo == nil {
```
Comment (mimowo) [inline pkg/visibility/api/rest/pending_workloads_cq.go:70]: Is the nil equivalent to non-existing ClusterQueue, or could this also happen when there are no pending workloads? I think the NotFound should only be returned when no ClusterQueue. I think it makes sense to have a test case for an existing cluster queue, but without any workloads.

**Review by mimowo — COMMENTED** [review]
/lgtm
/assign @alculquicondor @tenzen-y

File: pkg/visibility/api/rest/pending_workloads_cq_test.go [hunk 3]
```diff
@@ -298,15 +316,17 @@ func TestPendingWorkloadsInCQ(t *testing.T) {
 				manager.AddOrUpdateWorkload(w)
 			}
 
-			for _, cq := range tc.clusterQueues {
-				info, err := pendingWorkloadsInCqRest.Get(ctx, cq.Name, tc.queryParams)
+			for req, resp := range tc.wantResponse {
+				info, err := pendingWorkloadsInCqRest.Get(ctx, req.queueName, req.queryParams)
 				if err != nil {
```
Comment (alculquicondor) [inline pkg/visibility/api/rest/pending_workloads_cq_test.go:321]: We should always check the errors. For example, if `err` is nil, and `resp.wantErr` is nil, the test, as written, wouldn't report that the error is not found.

File: pkg/visibility/api/rest/pending_workloads_cq_test.go [hunk 4]
```diff
@@ -298,15 +316,17 @@ func TestPendingWorkloadsInCQ(t *testing.T) {
...
-					t.Fatal(err)
-				}
-				pendingWorkloadsInfo := info.(*visibility.PendingWorkloadsSummary)
-
-				if diff := cmp.Diff(tc.wantPendingWorkloads, pendingWorkloadsInfo.Items, cmpopts.EquateEmpty()); diff != "" {
-					t.Errorf("Pending workloads differ: (-want,+got):\n%s", diff)
+					if diff := cmp.Diff(resp.wantErr.Error(), err.Error()); diff != "" {
```
Comment (alculquicondor) [inline pkg/visibility/api/rest/pending_workloads_cq_test.go:322]: we don't compare strings https://github.com/kubernetes/community/blob/master/sig-scheduling/CONTRIBUTING.md#technical-and-style-guidelines

maybe you can have `wantErrMatch func(error) bool`, and you can pass things like `errors.IsNotFound`

File: pkg/visibility/api/rest/pending_workloads_cq_test.go [hunk 5]
```diff
@@ -298,14 +339,22 @@ func TestPendingWorkloadsInCQ(t *testing.T) {
...
+				// An error was not expected but it appeared
+				t.Error(err)
+			} else if err != nil {
+				// An error was expected and it appeared
+				if !tc.wantErrMatch(err) {
+					t.Errorf("Error differs: (-want,+got):\n%s", cmp.Diff(tc.wantResp.wantErr.Error(), err.Error()))
 				}
+			} else if tc.wantResp.wantErr != nil {
+				// An error was expected but it didn't appear
+				t.Errorf("Should have received an error: %v\n", tc.wantResp.wantErr)
```
Comment (alculquicondor) [inline pkg/visibility/api/rest/pending_workloads_cq_test.go:353]:
```
if tc.wantErrMatch != nil {
  if !tc.wantErrMatch(err) { t.Error() }
} else if err != nil {
    t.Error()
}

**Comment by alculquicondor** [comment]
/approve
/hold
for comment

**Comment by mimowo** [comment]
/hold cancel
The comment is addressed

**Comment by alculquicondor** [comment]
/release-note-edit
```release-note
NONE
```

**Pull request merged**